Basic clipboard support

ickshonpe opened the pull request:
# Objective

Add a platform-agnostic interface for interacting with the clipboard.

## Solution

New crate `bevy_clipboard` with a `ClipboardPlugin` that adds a `Clipboard` resource. The clipboard is accessed using the methods `fetch_text`, `fetch_image`, `set_text` and `set_image` on the `Clipboard` resource. `fetch_text` returns a `ClipboardRead` with a `poll_result` method that's used to get the actual value once it's ready.

The `windows` and `unix` implementations both use the `arboard` crate. On windows the `Clipboard` resource is a unit struct and a new arboard clipboard instance is created and dropped for each clipboard access. On unix targets the `Clipboard` resource holds a clipboard instance it reuses each time. On both targets the `fetch_*` and `set_*` methods work instantly.

On `wasm32` `Clipboard` is a unit struct. The `fetch_text` and `set_text` functions spawn async tasks. The task spawned by `fetch_text` updates the shared arc mutex option once the future is evaluated to get the clipboard text.  There is no image support on `wasm32`.

Everything seems to work but it feels like the design is a bit clumsy and not very idiomatic. I don't tend to do much asynchronous programming, maybe a reviewer can suggest an improved construction.

I also added an alternative `fetch_text_async` function for async access that returns a `Result<String, ClipboardError>` future.

### Notes
* Doesn't support android targets yet. 
* The wasm32 implementation doesn't support images. It's much more complicated and probably best to left to a follow up.

## Testing

The PR includes a basic example `clipboard` that can be used for testing.
The image display will only work if the image is already in the clipboard before the example starts.

ickshonpe committed:
Added new very basic clipboard crate.
Enabled with the `bevy_clipboard` feature.
Supports setting and getting text to and from the clipboard.
Only supports desktop targets.

ickshonpe committed:
Added new very basic clipboard crate.
Enabled with the `bevy_clipboard` feature.
Supports setting and getting text to and from the clipboard.
Only supports desktop targets.

ickshonpe committed:
Merge branch 'clipboard' of https://github.com/ickshonpe/bevy into clipboard

ickshonpe committed:
Revert changes to example

ickshonpe committed:
Merge branch 'main' into clipboard

alice-i-cecile commented:
Marked as X-Controversial due to a new crate, but I strongly think we need this as part of the engine in some form. Both for games and the editor!

ickshonpe committed:
ci fixes

ickshonpe committed:
Merge branch 'clipboard' of https://github.com/ickshonpe/bevy into clipboard

mockersf commented:
I don't like the crate for that which is pretty much just a resource wrapping arboard...

And not sure I like the fake resource that can't work on not supported platforms

ickshonpe committed:
removed `new` method

MrGVSV commented on crates/bevy_clipboard/src/dummy.rs:
```diff
@@ -0,0 +1,26 @@
+    ///
+    /// Returns error if clipboard is empty or contents are not UTF-8 text.
+    pub fn get_text(&mut self) -> Result<String, arboard::Error> {
+        Err(arboard::Error::ClipboardNotSupported)
```
Question: why can't `Clipboard` store the string itself? Then use `arboard` in addition when it is supported?

I guess that only works when setting directly through the resource. But might be an okay tradeoff

ickshonpe replied:
I thought about this but but seemed like it might be confusing to only support copy and paste internally.

MrGVSV replied:
What platforms are unsupported? It might be an okay tradeoff to still allow some amount of clipboard behavior

Henauxg replied:
`arboard` seems to support Linux, Mac and Windows. So there's at least Web that is unsupported, and many no_std targets.
I also feel like ultra basic support is still better than no support at all ? Maybe with a warning emitted if the plugin is enabled on an unsupported platform ?

ickshonpe replied:
I've added `wasm32` support now, it seems to work okay but needs testing.

All enabling the plugin on an unsupported platform does is add init a unit `Clipboard` resource with accessor functions that always return `ClipboardError::ClipboardNotSupported`, I don't know if it needs any warnings.

Henauxg commented on crates/bevy_clipboard/src/desktop.rs:
```diff
@@ -0,0 +1,30 @@
+        &mut self,
+        text: T,
+    ) -> Result<(), arboard::Error> {
+        arboard::Clipboard::new().and_then(|mut clipboard| clipboard.set_text(text))
```
I think dropping the `Clipboard` may not work on Linux ?
See [arboard's documentation](https://docs.rs/arboard/latest/arboard/struct.Clipboard.html#platform-specific-behavior) about platform-specific behaviors (and also issue [#88](https://github.com/1Password/arboard/issues/88) and PR [#89](https://github.com/1Password/arboard/pull/89) for more info)

ickshonpe replied:
Yeah seems like drop after use should only be done on windows.

ickshonpe replied:
Yeah I think on linux it needs to retain a clipboard instance.

Henauxg commented on crates/bevy_clipboard/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+# bevy
+bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
+arboard = "3.5.0"
```
We could disable the default `image-data` feature while we don't use it yet

ickshonpe commented:
> I don't like the crate for that which is pretty much just a resource wrapping arboard...

This PR is really trivial, but adding image and android and wasm support will add a lot more complexity. I'm being lazy here, I've got way too many things I working on atm so I was thinking if I just get desktop text clipboard support going someone else will pick this up and fill out the rest of the implementation. 

I'm not sure it's an important enough feature to deserve its own crate, but I'm not sure if it belongs in any of the exisiting crates. Maybe `bevy_window` but using the clipboard doesn't require a window.

> 
> And not sure I like the fake resource that can't work on not supported platforms

The idea is that you can only use the clipboard with a ResMut<Clipboard> to ensure that access is exclusive. Maybe there's a better way to implement it, I'm not sure.

ickshonpe committed:
Update crates/bevy_clipboard/Cargo.toml

Co-authored-by: Gilles Henaux <ghx_github_priv@fastmail.com>

ickshonpe committed:
* Added `ClipboardError` enum.
* On unix targets an instance of `arboard::Clipboard` is stored in the `Clipboard` resource.
* Removed the desktop and dummy modules.
* Made `arboard` dependency conditional.

ickshonpe committed:
Fixed incorrect error type.

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
Cleanup

ickshonpe committed:
Merge branch 'clipboard' of https://github.com/ickshonpe/bevy into clipboard

ickshonpe committed:
Removed unused variable

ickshonpe commented:
> For crates candidates, I've had a look and did not find much either. Doesn't fit in `bevy_text`, nor `bevy_input`. Could always force it in `bevy_utils` but that's no great either. There's maybe `bevy_app` which already has a `TerminalCtrlCHandlerPlugin` plugin using the `ctrlc` dependency, but I'm not sure that tying clipboard to `App` is a good choice.
> 
> There's a potential savior in `bevy_winit`. I've seen that clipboard handling _might_ be on the roadmap for 0.31 or later ? See [#162](https://github.com/rust-windowing/winit/issues/162) and [#2156](https://github.com/rust-windowing/winit/pull/2156) So it may make sense to snuck it into `bevy_winit`, while waiting for it to be available directly from winit ?

 It'd be ideal if winit just supported it and we didn't have to do anything but I don't like the idea of adding it to the winit crate before they have some built in support, it seems like a lie. And it might still need a separate solution when targeting wasm32?

ickshonpe committed:
added wasm32 support

ickshonpe committed:
Replaced `ClipboardText` with `ClipboardContents` enum.

ickshonpe committed:
Reverted example changes

ickshonpe committed:
Renamed `try_get` to `get_or_poll`, `get_text` to `fetch_text`

ickshonpe committed:
use Mutex from bevy_platform

ickshonpe committed:
Removed unused imports.

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
Added extra comment

ickshonpe committed:
Merge branch 'clipboard' of https://github.com/ickshonpe/bevy into clipboard

ickshonpe committed:
Removed futures depedency

alice-i-cecile reviewed (approved):
I have been convinced by @mweatherley that we should do this, largely to abstract over differences in platform support that currently aren't adequately handled on the web.

Because that's one of the core reasons to make or use this crate, it can't live in `bevy_widgets` (gameplay wants clipboard too), and there's no better home currently.

ickshonpe commented:
> I have been convinced by @mweatherley that we should do this, largely to abstract over differences in platform support that currently aren't adequately handled on the web.
> 
> Because that's one of the core reasons to make or use this crate, it can't live in `bevy_widgets` (gameplay wants clipboard too), and there's no better home currently.

Yep, especially since on windows clipboard access is synchronous and with arboard if you try to access the clipboard from two different systems simulateneously it'll panic. There needs to be some mechanism, like a `ResMut` resource, to enforce exclusive clipboard access. Users can't just import arboard themselves as there will be no way for them to tell if it's safe to access

ickshonpe committed:
Added `clipboard` example

ickshonpe committed:
Refactor to flatten the implementation.

Added `Clipboard` and `ClipboardContents` to the prelude.

ickshonpe committed:
Added message to discriminate unknown arboard errors.

ickshonpe committed:
renamed `get_or_poll` to `try_take`

ickshonpe committed:
Renamed `ClipboardContents` to `ClipboardRead`

ickshonpe committed:
Fixed grammar in comment

ickshonpe committed:
updated doc comments

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
Fixed cargo description

ickshonpe committed:
Fixed wasm32 imports

ickshonpe committed:
Cleaned up imports

ickshonpe committed:
Fixed spelling mistake

ickshonpe committed:
Edited comment.

ickshonpe committed:
convert string on wasm

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
Return  `Err(ClipboardError::ClipboardNotSupported)` when no clipboard support for platform.

ickshonpe committed:
Merge branch 'main' into clipboard

andrewzhurov committed:
Merge remote-tracking branch 'bevy/main' into clipboard

andrewzhurov committed:
0.16.0-dev -> 0.17.0-dev

andrewzhurov committed:
migrate clipboard example

ickshonpe committed:
Merge pull request #8 from andrewzhurov/clipboard

Clipboard bevy/main merge

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
Added `fetch_text_async` function

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
updated cargo dependencies

ickshonpe committed:
Raised Arboard version to 0.6.1, enabled its image-data feature.

Added `fetch_image` and `set_image` methods to `Clipboard`, which use access the clipboard to get images if present.

ClipboardRead now takes a generic type paramater, so it can handle both strings and images.

Updated the clipboard example to feature an image.

ickshonpe committed:
Added image feature gate

ickshonpe committed:
more fluency

ickshonpe committed:
more fluent refactors

ickshonpe committed:
Fix image to unwrap fetched image

ickshonpe committed:
build example pages

ickshonpe committed:
Merge branch 'main' into clipboard

ickshonpe committed:
fixed toml formatting

ickshonpe committed:
Merge branch 'clipboard' of https://github.com/ickshonpe/bevy into clipboard

ickshonpe committed:
update features

ickshonpe committed:
fix wasm errors

viridia commented on crates/bevy_clipboard/src/lib.rs:
```diff
@@ -0,0 +1,351 @@
+
+/// The clipboard prelude
+pub mod prelude {
+    pub use crate::{Clipboard, ClipboardRead};
```
plugin not part of the prelude?

ickshonpe committed:
tried to fix formatting

viridia commented on crates/bevy_clipboard/src/lib.rs:
```diff
@@ -0,0 +1,351 @@
+
+        #[cfg(not(any(unix, windows, target_arch = "wasm32")))]
+        {
+            ClipboardRead::Ready(Err(ClipboardError::ClipboardNotSupported))
```
What are your thoughts on having a fallback (perhaps controlled by a feature flag) that just stores the clip data locally. (Not in this PR however).

viridia commented on crates/bevy_clipboard/src/lib.rs:
```diff
@@ -0,0 +1,351 @@
+    /// Fetches UTF-8 text from the clipboard and returns it via a `ClipboardRead`.
+    ///
+    /// On Windows and Unix `ClipboardRead`s are completed instantly, on wasm32 the result is fetched asynchronously.
+    pub fn fetch_text(&mut self) -> ClipboardRead {
```
Something we should consider or document: when we talk about "text" this might mean "plain text", or it might mean a textual format such as JSON, TOML or BSN. All of these can be fetched with the same API, and the clipboard module should not be responsible for attempting to parse or reconcile these formats, as that is up to the client; however, if there is metadata (like MIME types) which allows us to know up front the advertised format of the clip, there should be an API to access that information. Perhaps that can be a field within the `ClipboardRead` type.

alice-i-cecile replied:
I would love to have better metadata about the format, but that should definitely be left to follow-up.

ickshonpe committed:
Added ClipboardPlugin to the prelude

alice-i-cecile commented:
The license checker here is wrong: BSL == [Boost Software License](https://www.boost.org/doc/user-guide/bsl.html), not "Business Software License". We should try to allow that crate (error-code) specifically if possible, since the other BSL should remain banned.

Zeophlite commented on Cargo.toml:
```diff
@@ -173,6 +173,15 @@ picking = ["bevy_picking", "mesh_picking", "sprite_picking", "ui_picking"]
 default_app = [
   "async_executor",
   "bevy_asset",
+  "bevy_audio",
```
This is probably a merge error

Zeophlite commented:
> The license checker here is wrong: BSL == [Boost Software License](https://www.boost.org/doc/user-guide/bsl.html), not "Business Software License". We should try to allow that crate (error-code) specifically if possible, since the other BSL should remain banned.

As per https://spdx.org/licenses/ , fix is to add `BSL-1.0` to `deny.toml` under `licenses.allow` - there's a blanket deny all for anything not specified.

"Business Source License 1.1" is `BUSL-1.1`

Zeophlite commented on examples/ui/clipboard.rs:
```diff
@@ -0,0 +1,227 @@
+//! This example demonstrates accessing the clipboard to retrieve and display text.
```
I originally had this as a reply, but figured it deserved it's own thread.

I feel like this example is confusing for how you'd actually use it.  You are expecting the clipboard content type, rather than reacting to it.

I think this example would be better if it can check the metadata of the Clipboard to see if it's an image, and if so display as such, otherwise display as text.

BUT looks like arboard doesn't support that :(

ickshonpe replied:
Yeah I made this example quickly without any thought just as proof it works. It's not necessary 
once the clipboard is hooked up to widgets that can use it.

alice-i-cecile replied:
I would also like to improve this example, but would much prefer to wait until follow-up.

alice-i-cecile committed:
Allow-list BSL-1.0 license needed

alice-i-cecile committed:
Merge branch 'main' into clipboard

alice-i-cecile committed:
Cleanup merge conflict problems

alice-i-cecile committed:
Remove bevy_clipboard from default features

This is both niche and security sensitive. Should not be on by default!

alice-i-cecile committed:
Clean up unrelated changes to Cargo.toml and remove accidental double newline

alice-i-cecile committed:
Write better documentation

alice-i-cecile committed:
Make arboard an optional, off-by-default feature

alice-i-cecile committed:
Improve docs on platform support

alice-i-cecile committed:
Improve docs on Clipboard

alice-i-cecile committed:
Simplify / unify windows arboard integration

alice-i-cecile committed:
Clipboard example is not wasm-safe

alice-i-cecile committed:
More docs work for `clipboard`

alice-i-cecile committed:
Improve design of clipboard features

alice-i-cecile committed:
Integrate with bevy_text

alice-i-cecile committed:
Remove asnyc unwrap in WASM

alice-i-cecile committed:
Defensive programming check for image text format

alice-i-cecile committed:
Fix incorrect self-referential feature

alice-i-cecile committed:
Impl Error + Display for ClipboardError

alice-i-cecile committed:
Remove unused fetch_text_async to clarify usage

alice-i-cecile committed:
Fix state machine bug in ClipboardRead: repeated polls would confusingly return Some(Err)

alice-i-cecile committed:
Improve feature flag setup

alice-i-cecile committed:
Log JsFuture errors in set_text

alice-i-cecile committed:
cargo run -p build-templated-pages -- update features

alice-i-cecile committed:
ci plz

alice-i-cecile committed:
Add note about deadlock risk on Windows

alice-i-cecile committed:
Add note to text_input example

alice-i-cecile committed:
Typo

alice-i-cecile committed:
Collapsible if

alice-i-cecile commented:
@ickshonpe, I'm running into some difficulties with getting image support working again. `cargo run --example clipboard --features="system_clipboard"` runs fine. However, it doesn't seem to detect my image properly (using PopOS). Can you take a look?

I am also 100% okay with cutting image support from this initial PR frankly.

viridia commented on crates/bevy_text/src/text_edit.rs:
```diff
@@ -192,43 +192,77 @@ impl TextEdit {
+                        // On wasm, set_text fires-and-forgets: the browser clipboard write is
+                        // async and we cannot confirm it succeeded before deleting. To avoid
+                        // silently discarding the selection, we copy only.
+                        bevy_log::warn!(
```
Not sure I agree with this decision.

In the case where a cut action fails to copy to the clipboard (which happens to me from time to time even on desktop OSs), my muscle memory response is to go back to the text input, hit undo, and try to cut again. So as long as we have undo support (which we will), then nothing should be "lost".

It would be strange if "cut" didn't appear to do anything visible - I expect that the user's response will be to try cutting again.

The relevant question is, how likely is this failure to occur?

alice-i-cecile committed:
Improve wording on clipboard operation features

alice-i-cecile committed:
Still delete text when cutting on wasm

alice-i-cecile committed:
Missed doc link

alice-i-cecile committed:
Remove wasm branching on paste for now

alice-i-cecile committed:
Leave TODO with thoughts

alice-i-cecile committed:
Handle async nature of pasting properly (and deduplicate text filter checks)

alice-i-cecile committed:
Typo

alice-i-cecile committed:
Make pending_paste pub to preserve ..default initialization

alice-i-cecile commented:
Okay, I think this is finally ready! 

1. The image support functions in general, but for some reason does not like my screenshot snip tool on PopOS.
2. System clipboard support is now fully operational, including on web. It needed some fanciness to support polling properly, but I'm really quite pleased with the final architecture.

@ickshonpe, this is ready to merge once my changes have your approval.

alice-i-cecile committed:
Fix missed import on wasm

alice-i-cecile committed:
Merge branch 'main' into clipboard

ickshonpe commented on crates/bevy_internal/Cargo.toml:
```diff
@@ -459,6 +459,15 @@ gamepad = ["bevy_input/gamepad", "bevy_input_focus?/gamepad"]
+# Clipboard support
+bevy_clipboard = ["dep:bevy_clipboard"]
+system_clipboard = [
+  "dep:bevy_clipboard",
+  "bevy_clipboard/system_clipboard",
+  "bevy_clipboard/image",
+  "bevy_text?/system_clipboard",
+]
```
This tries to enable the image feature on wasm32, doesn't it?
```
cargo check --features="system_clipboard" --target wasm32-unknown-unknown
```
fails

alice-i-cecile committed:
Remove clipboard example used for testing

alice-i-cecile committed:
Split apart image clipboard support b/c broken on WASM

ickshonpe commented on crates/bevy_text/src/text_editable.rs:
```diff
@@ -144,6 +148,7 @@ impl Default for EditableText {
             // Defaults selected to match `Text::default()`
             editor: PlainEditor::new(100.),
             pending_edits: Vec::new(),
+            pending_paste: None,
```
In `bevy_ui_text_input` it queues a `PasteDeferred(ClipboardRead)` in the editor actions buffer so there isn't a need to store it on a field. Then if the result is still pending it doesn't process any more actions and leaves the PasteDeferred on top of the queue to be processed again the next frame. The pending field approach seems a bit simpler, but we might want to switch to the other approach later as it seems unpleasant having `bevy_text` depend on `bevy_clipboard` and there might be other reasons we'd want to handle cut and paste at the widget level (pasting images inline into a text input maybe?).

ickshonpe replied:
I can't remember what they were right now, but I'm certain there were other advantages to managing the `TextEdit` queue at the widget level, not at a lower level from `bevy_text`. It's out of scope for this PR though, the `pending_paste` design is fine for now.

alice-i-cecile replied:
Yeah, happy to explore that in follow-up. I agree that bevy_text -> bevy_clipboard is a bit unpleasant as a dependency.

alice-i-cecile committed:
CI formatting

alice-i-cecile committed:
cargo run -p build-templated-pages -- update features

ickshonpe commented:
can't approve my own pull request, but 👍

viridia reviewed (approved):
Re-approving
